Feature/#12 Error Handling + 기존 에러처리 되어있지 않던 뷰에 에러처리 적용(Error Alert)

minsson opened the pull request:
@jcrescent61 

안녕하세요, 엘렌!
오랜만에 보내는 PR이네요! 마무리까지 힘내봅시다! 😁

## 구현 내용
- 모든 뷰에서 편리하게 사용할 수 있는 Error Alert Modifier를 구현했습니다.
    ```swift
    func errorAlert<E: LocalizedError>(error: Binding<E?>, buttonTitle: String = "알겠어요 ☺️") -> some View
    ```
    - 뷰모델에서 LocalizedError 프로토콜을 채택한 타입인 Published Property를 들고 있고, 뷰에서 이 프로퍼티를 바인딩해 errorAlert 모디파이에 넣어주면, 바인딩 된 error 프로퍼티가 업데이트될 때마다 Error Alert를 띄워주는 방식입니다.
- 앱 전반적으로 사용할 에러 프로토콜과 모델을 구현했습니다.
    - GalpiErrorProtocol, GalpiError 입니다.
    - LocalizedError 프로토콜을 채택해 유저에게 보여줄 errorDescription, recoverySuggestion을 갖고 있습니다.
    - 다만 완벽한 에러처리를 위해서는 모든 에러케이스에 대해 구체적인 errorDescription, recoverySuggestion를 Alert로 띄워줘야겠지만, 현재 단계에서는 일종의 오버엔지니어링이라고 생각해 범용적인 에러메시지를 사용하고 있습니다. 만약 추후 필요하다면 errorDescription과 recoverySuggestion의 String들만 바꿔주면 됩니다.
    - 따라서, 현재의 Error Alert 내용은 아래와 같습니다.
        - Alert의 타이틀(errorDescription): '에러가 발생했어요 😿'로 통일
       - Alert의 내용(recoverySuggestion): '다시 해봐도 안 되면 개발자에게 에러 코드를 말씀해주세요. 빠르게 해결해드릴게요! + (에러코드: @@@)'
            | Error Alert 예시 이미지 |
            | :-: | 
            | <img src = "https://user-images.githubusercontent.com/96630194/225217758-6587689c-e291-4b58-b6b5-622b683ffae0.png" height="500px"> | 
- 필요하다면 각 객체마다 에러 타입을 구현할 수 있습니다. LoginView에서 FirebaseAuth를 통해 애플 로그인을 하는 것과 관련, GalpiErrorProtocol을 채택한 ServerAuthError와 SocialLoginError 타입을 구현해두었습니다. 
    - 이런 에러들은 컴플리션 핸들러와 Result 타입, 그리고 enum의 연관값을 이용해 더 작은 집합의 에러에서 더 큰 집합의 에러로 넘어가, 최종적으로 View Model에 도달해 Error Alert로 띄워집니다.
    - 예를 들어, 애플 로그인에서 `SocialLoginError.canceled` 에러가 다른 객체로 넘어가는 방식은 아래와 같습니다.
        - AppleLoginManager에서 `SocialLoginError.canceled` 발생 -> FirebaseAuthService에 `SocialLoginError.canceled` 전달
        - FirebaseAuthService에서는 AppleLoginManager에서 온 에러를 ServerAuthError의 socialLoginError(SocialLoginError)로 받음 -> LoginViewModel에 `ServerAuthError.socialLoginError(SocialLoginError.canceled)` 전달
        - LoginViewModel에서는 FirebaseAuthService에서 온 에러를 GalpiError의 serverAuthError(ServerAuthError)로 받음 -> self.error에 `GalpiError.serverAuthError(ServerAuthError.socialLoginError(SocialLoginError.canceled)` 할당
        - LoginView에서 `.errorAlert(error: $viewModel.error)`를 통해 LoginViewModel 내 error 프로퍼티의 업데이트 여부  확인 -> 업데이트 되면 Error Alert를 띄움

## 고민했던 점
- 위와 같이 에러처리를 할 때의 장단점
-  저는 장점이 더 크다고 생각해 이렇게 구현했는데, 엘렌은 어떻게 생각하시는지 궁금합니다.
    - 장점:
        - 각 뷰에서 하나의 errorAlert 모디파이어, 각 뷰모델에서 하나의 error Published Property만 만들어주면 되어 편리함
        - 에러처리의 구조가 짜여있으므로, 앱이 확장되더라도 구조에 맞춰 에러 타입과 케이스만 정의해주면 되어 편리함
    - 단점:
        - 에러 타입의 수가 늘어남
        - 더 작은 에러에서 더 큰 에러로 넘어가며 전달되는 구조가 복잡하게 느껴질 수 있음  

## linked issue
- #12

minsson committed:
#12 feat: 제네릭하게 사용 가능한 errorAlert 모디파이어 구현

- LocalizedError 프로토콜을 준수하는 제네릭 에러 타입을 받아 alert를 띄워줍니다.
- 뷰 모델의 error 프로퍼티와 바인딩해, error 프로퍼티가 업데이트 될 때마다 errorAlert 모디파이어가 실행되어 Alert가 띄워집니다.
- alert 메서드의 isPresented에 들어가는 .constant(localizedError != nil)의 경우 Binding<true> 또는 Binding<false>가 상수로 들어갑니다.
    - 상수로 들어가지만, 여기서 사용하는 alert 메서드의 공식문서에 따르면 아무 버튼이나 누를 시 시스템 차원에서 Binding<false>로 변경하고, Alert도 dismiss 해줍니다.

minsson committed:
#12 feat: GalpiErrorProtocol 정의

minsson committed:
#12 feat: SocialLoginError 정의

minsson committed:
#12 feat: SocialLoginError 케이스 추가

minsson committed:
#12 feat: SocialLoginError에 개발자를 위한 errorReason 추가

minsson committed:
#12 feat: GalpiErrorProtocol에 errorReason 요구사항 추가

minsson committed:
#12 refactor: SocialLoginManagerProtocol의 에러 타입 변경

- 기존에 Error 타입으로 임시구현했던 코드를 실제 에러 핸들링을 위해 수정했습니다.

minsson committed:
#12 feat: AppleLoginManager에 에러 처리 코드 적용

- guard let 바인딩에서 nil이 나오는 경우는 거의 없을 것이라고 생각하지만, 우선 에러 처리는 해두었습니다.

- ASAuthorizationControllerDelegate의 델리게이트 메서드를 통해 어떤 에러가 있었는지 전달 받습니다.
    - authorizationController(controller: didCompleteWithError:) 메서드입니다.
    - 전달 받은 error를 ASAuthorizationError로 캐스팅해 다양한 에러 케이스를 전달 받을 수 있습니다.
    - 하지만 현재 단계에서 그렇게 자세하고 구체적인 에러 핸들링이 필요한지 여부가 불투명해 두 케이스로만 나누어두었습니다.
        - 유저가 직접 로그인을 취소한 경우와, 그 외의 경우입니다.

minsson committed:
#12 feat: ServerAuthError 정의

- case 중 socialLoginError(SocialLoginError)
    - FirebaseAuthService에서 socialLoginManager의 signIn 메서드를 실행할 때 Result 타입으로 SocialLoginError를 전달 받아 LoginViewModel로 넘겨주기 위해 사용합니다.

minsson committed:
#12 refactor: ServerAuthServiceProtocol의 에러 타입 변경

- 기존에 Error 타입으로 임시구현했던 코드를 실제 에러 핸들링을 위해 수정했습니다.

minsson committed:
#12 feat: FirebaseAuthService에 에러 처리 코드 적용

minsson committed:
#12 feat: 모든 에러를 전달 받아 Alert를 띄울 때 쓰일 GalpiError 정의

- 뷰모델에서 갖는 error 프로퍼티의 타입으로, View의 errorAlert 모디파이어에 들어갈 수 있습니다.

minsson committed:
#12 feat: LoginViewModel에 에러 처리 코드 적용

minsson committed:
#12 feat: LoginView에서 에러 발생시 Alert 띄우는 기능 구현 완료

minsson committed:
#12 feat: 에러 Alert에 사용되는 메시지 내용 수정

- 이모지가 너무 많이 쓰이는 것 같아 제거했습니다.

minsson committed:
#12 refactor: ServerAuthError의 에러코드 수정

- FirebaseAuth가 아니라 ServerAuth이므로 이에 맞춰 S로 수정했습니다.

minsson commented on galpi/galpi/Domain/Common/Error/GalpiErrorProtocol.swift:
```diff
@@ -0,0 +1,32 @@
+protocol GalpiErrorProtocol: LocalizedError {
+    
+    var code: String { get }
+    var errorReason: String { get }
```
각 에러 케이스에 대해 에러 원인이 무엇인지에 대한 설명이 들어갑니다.

유저에게 보여질 내용은 아니며, 우리가 에러 케이스 이름만으로는 정확히 어떤 에러이고, 원인이 무엇인지 파악하기 어려울 수 있다고 생각해 일종의 주석 용도로 만들어두었습니다. 

그렇지만, 주석 용도라면 차라리 그냥 주석을 작성하는 게 낫겠다는 생각도 듭니다. 
엘렌은 어떻게 생각하시나요?

jcrescent61 replied:
P1: 유저에게 보일 Alert 메세지에는 `에러가 발생했어요`를 띄워주고
로그에서는 `Network Error + Code`를 띄워주는 것이 좋다고 생각합니다.

읽을거리가 많아지고 에러가 발생하는 시점을 log에서 확인할 수 없는 점에서 주석처리는 좋지 않다고 생각합니다.

minsson replied:
의견 감사합니다. 공감합니다 ☺️

minsson commented on galpi/galpi/Domain/Common/Error/GalpiErrorProtocol.swift:
```diff
@@ -0,0 +1,32 @@
+
+extension GalpiErrorProtocol {
+    
+    var errorDescription: String? {
+        switch self {
+        default:
+            return "에러가 발생했어요 😿"
+        }
+    }
```
PR 본문에 자세히 써놨듯이, Error Alert의 타이틀이 됩니다.

현재 단계에서는 Error Alert에 매번 '로그인 실패', '무슨무슨 오류' 등으로 각기 다른 타이틀을 달 필요성이 느껴지지 않아 default로 "에러가 발생했어요 😿"를 반환하고 있습니다.

jcrescent61 replied:
네 좋습니다!

minsson commented on galpi/galpi/Domain/Common/Error/GalpiErrorProtocol.swift:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+    
+    var recoverySuggestion: String? {
+        switch self {
+        default:
+            return "다시 해봐도 안 되면 개발자에게 에러 코드를 말씀해주세요. 빠르게 해결해드릴게요! \n" + "(에러코드: \(code))"
+        }
+    }
```
Error Alert가 띄워진다면 일반적으로는 앱 자체에서 발생하는 에러일 것 같습니다. 각 에러 케이스마다 다른 에러 코드를 부여해, 유저가 우리에게 제보할 수 있도록 해두었습니다.

다만, 로그인 도중 유저의 네트워크가 끊기는 등, 정말로 유저가 조치해야할 상황이 발생할 수도 있습니다. 이럴 때는 네트워크에 잘 연결되어 있는지 확인해달라는 등의 구체적인 에러메시지를 띄울 수 있을 것 같습니다. 

아직은 이런 케이스를 구현하지 않았습니다. 구현한다면 출시 전후 리팩토링 과정에서 구현하면 될 것 같습니다.

jcrescent61 replied:
P1: errorReason과 recoverySuggestion의 차이를 잘 모르겠습니다. 둘 다 필요한가요? 하나만 사용할 수 있지 않을까요?

minsson replied:
errorReason은 개발자가 보기 위한 에러 원인이고, recoverySuggestion은 유저에게 Alert를 띄울 때 Alert의 본문으로 띄우기 위한 프로퍼티입니다. 때문에 역할이 다르다고 생각합니다.

다만, errorReason의 경우 수정 사항이 있습니다.
아래 이미지를 보시면 LocalizedError에 failureReason이 있다는 걸 볼 수 있습니다. 제가 생각했던 errorReason과 같은 내용이라고 판단해 errorReason을 failureReason으로 수정했습니다. 
(참고 커밋: 70db6530bbd8f44820fa2702751aa202b734c016)

<img width="1020" alt="image" src="https://user-images.githubusercontent.com/96630194/225891278-b1ff11ba-3ce7-4057-833f-4dbd645eb822.png">

minsson committed:
#12 refactor: 각 에러 타입의 에러 코드 네이밍 수정

- SocialLogin의 경우 Apple Login 뿐만 아니라 Kakao, Google 등 다양한 로그인이 생길 수 있습니다.
    - 따라서 Apple의 A가 아니라, SocialLogin의 S로 수정했습니다.
- ServerLogin의 경우 이전 커밋에서 S로 수정했지만, 위의 SocialLogin의 S와 겹치게 되었습니다.
    - FirebaseAuth에서 다른 것으로 바뀔 가능성은 희박하다고 생각해, S에서 F로 다시 수정했습니다.

minsson commented on galpi/galpi/Domain/Common/Error/SocialLoginError.swift:
```diff
@@ -0,0 +1,41 @@
+    case identityToken
+    case idTokenString
+    
+    var code: String {
+        switch self {
+        case .canceled:
+            return "S100"
+        case .internalError:
+            return "S200"
+        case .identityToken:
+            return "S300"
+        case .idTokenString:
+            return "S400"
+        }
+    }
```
유저에게 표시될 에러코드입니다. 
SocialLogin이므로 S를 앞에 붙였습니다.
100, 200 등의 숫자 자체에는 의미가 없으며, 각 에러케이스를 구분하기 위한 용도로 임의로 작성되었습니다.

jcrescent61 replied:
네 좋습니다!

minsson commented on galpi/galpi/Domain/Common/Error/ServerAuthError.swift:
```diff
@@ -0,0 +1,31 @@
+    case internalError
+    case socialLoginError(SocialLoginError)
+    
+    var code: String {
+        switch self {
+        case .internalError:
+            return "F100"
+        case .socialLoginError(let error):
+            return error.code
+        }
+    }
```
유저에게 표시될 에러코드입니다. 
ServerAuth이므로 S를 앞에 붙이면 좋겠지만, SocialLogin과 겹쳐 S를 쓸 수 없었습니다.
FirebaseAuth에서 다른 것으로 바뀔 가능성은 희박하다고 생각해 F로 해두었습니다.
100, 200 등의 숫자 자체에는 의미가 없으며, 각 에러케이스를 구분하기 위한 용도로 임의로 작성되었습니다.

jcrescent61 replied:
P1: 에러 코드 기준을 정할땐 회의를 통해 함께 정해보았으면 좋았겠다 싶은데요 :)
알파벳이 한 개만 들어가는 것이 좋지 않다고 생각합니다.
중복 알파벳이 생겨나는 경우를 대비해서 최소 두개 알파벳이 들어가는 것이 좋을 것 같다는 생각이 듭니다!

minsson replied:
말씀하신 내용에 공감합니다. 알파벳 두 글자로 수정했습니다!
(참고 커밋: 1908d0c4f7b60f1c15aafe1eea208782664b5841)

minsson commented on galpi/galpi/Extension/View+Extension.swift:
```diff
@@ -0,0 +1,21 @@
+
+import SwiftUI
+
+extension View {
+    func errorAlert<E: LocalizedError>(error: Binding<E?>, buttonTitle: String = "알겠어요 ☺️") -> some View {
+        let localizedError = error.wrappedValue
+        return alert(isPresented: .constant(localizedError != nil), error: localizedError) { _ in
+            Button(buttonTitle) {
+                error.wrappedValue = nil
+            }
+        } message: { error in
+            Text(error.recoverySuggestion ?? "")
+        }
+    }
+}
```
- LocalizedError 프로토콜을 준수하는 제네릭 에러 타입을 받아 alert를 띄워줍니다.
- 뷰 모델의 error 프로퍼티와 바인딩해, error 프로퍼티가 업데이트 될 때마다 errorAlert 모디파이어가 실행되어 Alert가 띄워집니다.
- alert 메서드의 isPresented에 들어가는 .constant(localizedError != nil)의 경우 Binding<true> 또는 Binding<false>가 상수로 들어갑니다.
    - 상수로 들어가지만, 여기서 사용하는 alert 메서드의 공식문서에 따르면 아무 버튼이나 누를 시 시스템 차원에서 Binding<false>로 변경하고, Alert도 dismiss 해줍니다.

jcrescent61 replied:
저번에 설명해주셨던 내용인 것 같은데 다시 보니 이해가 잘 가지 않는 군요! 스크럼때 이해시켜 주시길 바랍니다🤣

minsson replied:
알겠습니다 😁

minsson reviewed (commented):
코드 리뷰하실 때 참고하실 수 있도록 코드 일부에 미리 코멘트를 남겨둡니다. ☺️

jcrescent61 commented on galpi/galpi/Domain/Common/Error/SocialLoginError.swift:
```diff
@@ -0,0 +1,41 @@
+        case .identityToken:
+            return "identity token의 fetch 실패"
+        case .idTokenString:
+            return "data를 token string으로 직렬화하는 것 실패"
```
P2: 위의 에러 메세지 관련 코멘트와 비슷한 내용이 될 것 같습니다.
이런 내용이 로그에 들어가는 것은 좋다고 생각합니다. 다만 errorReason 이라는 네이밍 보다 description이 더 좋을 것 같습니다.

jcrescent61 commented on galpi/galpi/Services/OAuth/FirebaseAuthService.swift:
```diff
@@ -31,16 +31,16 @@ final class FirebaseAuthService: ServerAuthServiceProtocol {
                 
                 Auth.auth().signIn(with: firebaseCredential) { (authResult, error) in
-                    if let error = error {
-                        print(error)
```
Q: 로그에 error 메세지를 띄우지 않으시는 이유가 있으실까요? 궁금합니다!

minsson replied:
에러 처리를 제대로 하는 것은 이번이 처음이라, 프린트로 로그를 찍는 것을 염두에 두지 못했습니다 😅
현재는 아래와 같이 구현했습니다. 로그의 경우 아래의 에러가 Result 타입으로 Completion Handler를 타고 뷰모델까지 올라가 프린트 되게 됩니다.

```swift
Auth.auth().signIn(with: firebaseCredential) { (authResult, error) in
    guard error == nil else {
        completion(.failure(ServerAuthError.internalError))
        return
    }
    
    completion(.success(true))
}
```

minsson replied:
프린트로 로그를 찍는 것을 염두에 두지 못했네요 😅
로그의 경우 아래의 에러가 Result 타입으로 Completion Handler를 타고 뷰모델까지 올라가 프린트 되도록 수정했습니다. (참고 커밋: 3c42db3e858b9b864ceacc589173d760d472ed6e)

```swift
Auth.auth().signIn(with: firebaseCredential) { (authResult, error) in
    guard error == nil else {
        completion(.failure(ServerAuthError.custom(error?.localizedDescription ?? "")))
        return
    }
    
    completion(.success(true))
}
```

minsson replied:
프린트로 로그를 찍는 것을 염두에 두지 못했네요. 아래의 에러가 Result 타입으로 Completion Handler를 타고 뷰모델까지 올라가 프린트 되도록 수정했습니다. (참고 커밋: 3c42db3e858b9b864ceacc589173d760d472ed6e)

```swift
Auth.auth().signIn(with: firebaseCredential) { (authResult, error) in
    guard error == nil else {
        completion(.failure(ServerAuthError.custom(error?.localizedDescription ?? "")))
        return
    }
    
    completion(.success(true))
}
```

minsson replied:
프린트로 로그를 찍는 것을 염두에 두지 못했네요. 에러가 Result 타입으로 Completion Handler를 타고 뷰모델까지 올라가 프린트 되도록 아래와 같이 수정했습니다. (참고 커밋: 3c42db3e858b9b864ceacc589173d760d472ed6e)

```swift
Auth.auth().signIn(with: firebaseCredential) { (authResult, error) in
    guard error == nil else {
        completion(.failure(ServerAuthError.custom(error?.localizedDescription ?? "")))
        return
    }
    
    completion(.success(true))
}
```

jcrescent61 commented on galpi/galpi/Services/OAuth/FirebaseAuthService.swift:
```diff
@@ -10,7 +10,7 @@ import FirebaseAuth
     
     var isSignedIn: Bool { get }
-    func signIn(with socialLogin: SocialLogin, completion: @escaping (Result<Bool, Error>) -> Void)
+    func signIn(with socialLogin: SocialLogin, completion: @escaping (Result<Bool, ServerAuthError>) -> Void)
```
그냥 Error가 아닌 구체적으로 어떤 에러가 반환되는지 알 수 있어서 좋습니다!

jcrescent61 reviewed (commented):
민쏜 안녕하세요 고생많으셨습니다! 에러처리 어떠셨나요 재밌으셨나요? ㅎㅎ
이번 리뷰는 스크럼때 할 이야기가 많을 것 같습니다!

**주제**
- 유저에게 보여질 에러메세지와 log에 남겨야할 메세지를 어떻게 나눌 것인가?
- 현재 description, errorReason, recoverySuggestion 등 메세지의 종류가 너무 많지 않은가?

민쏜이 구현하신 로직대로 사용하신다면 유저에게 보여질 에러메세지, 로그에 남겨져야할 메세지, 에러코드 이정도로 나누어 볼 수 있을 것 같습니다. `recoverySuggestion`같은 경우에는 필요성을 잘 모르겠습니다...!!!

minsson committed:
#12 refactor: 각 에러 타입의 에러 코드 알파벳을 두 글자로 변경

- 알파벳 한 글자로는 겹치는 글자가 많을 것으로 예상되어, 두 글자로 변경했습니다.

minsson committed:
#12 refactor: errorReason을 LocalizedError의 failureReason으로 대체

- 기존에 작성했던 errorReason는 LocalizedError 프로토콜에 포함된 failureReason과 같은 의미입니다.
- 공식 API에 포함된 프로퍼티가 있으므로, 혼동을 줄이기 위해 기존의 errorReason을 failureReason으로 대체했습니다.

minsson committed:
#12 feat: 모든 세세한 에러 케이스를 구현할 필요가 없도록 커스텀 에러 케이스 구현

- 언래핑 실패 등 세부적인 에러 케이스가 많고, 이런 경우의 에러 처리는 비용에 비해 효용이 적다고 생각합니다.
- 따라서 우선은 별도의 에러 케이스 정의 없이도 에러 메시지를 호출하는 쪽에서 전달할 수 있도록 custom 에러 케이스를 만들었습니다.
- 추후 에러케이스 추가가 필요하다면 각 구체 타입에서 에러 케이스만 추가하면 되어 확장에도 지장이 없다고 생각했습니다.

minsson committed:
#12 refactor: 기존 일부 에러 케이스를 custom 에러 케이스로 변경

- 유의미하지 않다고 생각되는 에러케이스이므로 custom 에러 케이스로 변경했습니다.

minsson committed:
#12 feat: 소셜 로그인 실패 시 에러 로그가 출력되는 기능 구현

- 유저를 위한 Alert 뿐만 아니라, 개발자도 에러 원인을 파악하기 쉽도록 에러 로그를 출력합니다.

minsson committed:
#12 refactor: FirebaseAuth에서 내려주는 에러 내용이 로그에 담기도록 수정

- 에러가 Result 타입으로 Completion Handler를 타고 뷰모델까지 올라가, 에러 로그로 프린트 됩니다.

minsson committed:
#12 refactor: 애플 로그인에서 내려주는 에러 내용이 로그에 담기도록 수정

- ASAuthorizationControllerDelegate에서 델리게이트 메서드를 통해 내려주는 에러가 Result 타입으로 Completion Handler를 타고 뷰모델까지 올라가, 에러 로그로 프린트 되도록 수정했습니다.

minsson committed:
#12 refactor: custom 에러 케이스로 인해 사용되지 않는 internalError 케이스 제거

- 기존에 나누기 모호한 에러의 경우 internalError 케이스를 사용하고 있었는데, custom 에러 케이스를 구현하게 되며 지금은 필요하지 않게 되었습니다.

minsson committed:
#12 refactor: SocialLoginError 실패 이유를 더 자세한 내용으로 수정

- 기존 내용은 무슨 인증 절차인지 명확하지 않아 수정했습니다.
